File: samples/dotnet/OnnxTokenizer.Tests/EmbeddingComparisonTests.cs
using Newtonsoft.Json;
using OnnxTokenizer.Sample;

namespace OnnxTokenizer.Tests;

public class EmbeddingComparisonTests : IDisposable
{
    private readonly OnnxEmbeddingGenerator _embeddingGenerator;
    private readonly Dictionary<string, float[]> _referenceEmbeddings;

    public EmbeddingComparisonTests()
    {
        string projectDir = Path.GetFullPath(Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "..", "..", ".."));
        string repoDir = Path.GetFullPath(Path.Combine(projectDir, "..", "..", ".."));
        string onnxDir = Path.Combine(repoDir, "onnx");
        string tokenizerPath = Path.Combine(onnxDir, "tokenizer.onnx");
        string modelPath = Path.Combine(onnxDir, "model.onnx");
        string referenceFile = Path.Combine(onnxDir, "reference_embeddings.json");

        if (!File.Exists(tokenizerPath))
            throw new FileNotFoundException($"Tokenizer file not found at {tokenizerPath}");
        if (!File.Exists(modelPath))
            throw new FileNotFoundException($"Model file not found at {modelPath}");
        if (!File.Exists(referenceFile))
            throw new FileNotFoundException($"Reference embeddings file not found at {referenceFile}");

        _embeddingGenerator = new OnnxEmbeddingGenerator(tokenizerPath, modelPath);

        string jsonContent = File.ReadAllText(referenceFile);
        _referenceEmbeddings = JsonConvert.DeserializeObject<Dictionary<string, float[]>>(jsonContent)!;
    }

    [Fact]
    public void NetEmbeddings_ShouldMatchPythonEmbeddings()
    {
        foreach (var entry in _referenceEmbeddings)
        {
            string text = entry.Key;
            float[] referenceEmbedding = entry.Value;

            // Generate embedding using .NET implementation
            float[] dotNetEmbedding = _embeddingGenerator.GenerateEmbedding(text);

            // Calculate cosine similarity
            double similarity = CalculateCosineSimilarity(dotNetEmbedding, referenceEmbedding);

            // Assert high similarity (practically identical)
            Assert.True(similarity > 0.9999,
                $"Expected similarity > 0.9999, but got {similarity} for text: {text}");
        }
    }

    private static double CalculateCosineSimilarity(float[] vectorA, float[] vectorB)
    {
        if (vectorA.Length != vectorB.Length)
            throw new ArgumentException("Vectors must be of the same length");

        double dotProduct = 0;
        double normA = 0;
        double normB = 0;

        for (int i = 0; i < vectorA.Length; i++)
        {
            dotProduct += vectorA[i] * vectorB[i];
            normA += vectorA[i] * vectorA[i];
            normB += vectorB[i] * vectorB[i];
        }

        return dotProduct / (Math.Sqrt(normA) * Math.Sqrt(normB));
    }

    public void Dispose()
    {
        _embeddingGenerator?.Dispose();
    }
}

File: samples/java/src/test/java/com/yunikosoftware/onnxtokenizer/EmbeddingComparisonTests.java
package com.yunikosoftware.onnxtokenizer;

import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.FileNotFoundException;
import java.io.FileReader;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Map;

import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.AfterAll;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

public class EmbeddingComparisonTests {
    private static OnnxEmbeddingGenerator embeddingGenerator;
    private static Map<String, float[]> referenceEmbeddings;

    @BeforeAll
    public static void setup() throws Exception {
        Path tokenizerPath = resolveModelPath("tokenizer.onnx");
        Path modelPath = resolveModelPath("model.onnx");
        Path referenceFile = resolveModelPath("reference_embeddings.json");

        if (!tokenizerPath.toFile().exists())
            throw new FileNotFoundException("Tokenizer file not found at " + tokenizerPath);
        if (!modelPath.toFile().exists())
            throw new FileNotFoundException("Model file not found at " + modelPath);
        if (!referenceFile.toFile().exists())
            throw new FileNotFoundException("Reference embeddings file not found at " + referenceFile);

        embeddingGenerator = new OnnxEmbeddingGenerator(tokenizerPath.toString(), modelPath.toString());

        ObjectMapper mapper = new ObjectMapper();
        referenceEmbeddings = mapper.readValue(
            new FileReader(referenceFile.toFile()),
            new TypeReference<Map<String, float[]>>() {}
        );
    }

    @Test
    public void javaEmbeddings_ShouldMatchPythonEmbeddings() throws Exception {
        for (Map.Entry<String, float[]> entry : referenceEmbeddings.entrySet()) {
            String text = entry.getKey();
            float[] referenceEmbedding = entry.getValue();

            float[] javaEmbedding = embeddingGenerator.generateEmbedding(text);

            double similarity = calculateCosineSimilarity(javaEmbedding, referenceEmbedding);

            assertTrue(similarity > 0.9999,
                "Expected similarity > 0.9999, but got " + similarity + " for text: " + text);
        }
    }

    private static double calculateCosineSimilarity(float[] vectorA, float[] vectorB) {
        if (vectorA.length != vectorB.length)
            throw new IllegalArgumentException("Vectors must be of the same length");

        double dotProduct = 0;
        double normA = 0;
        double normB = 0;

        for (int i = 0; i < vectorA.length; i++) {
            dotProduct += vectorA[i] * vectorB[i];
            normA += vectorA[i] * vectorA[i];
            normB += vectorB[i] * vectorB[i];
        }

        return dotProduct / (Math.sqrt(normA) * Math.sqrt(normB));
    }

    private static Path resolveModelPath(String filename) throws FileNotFoundException {
        Path currentDir = Paths.get("").toAbsolutePath();

        for (int i = 0; i < 5; i++) {
            Path onnxDir = currentDir.resolve("onnx");

            if (onnxDir.toFile().exists() && onnxDir.toFile().isDirectory()) {
                return onnxDir.resolve(filename);
            }

            Path parent = currentDir.getParent();
            if (parent == null)
                break;
            currentDir = parent;
        }

        throw new FileNotFoundException("Could not locate '" + filename + "' in the 'onnx' directory");
    }

    @AfterAll
    public static void cleanup() throws Exception {
        if (embeddingGenerator != null) {
            embeddingGenerator.close();
        }
    }
}